src/function/gds/awsp_paths.cpp

```diff
@@ -0,0 +1,179 @@
+#include "binder/expression/node_expression.h"
+#include "common/exception/interrupt.h"
+#include "function/gds/auxiliary_state/path_auxiliary_state.h"
+#include "function/gds/gds_function_collection.h"
+#include "function/gds/rec_joins.h"
+#include "main/client_context.h"
+#include "processor/execution_context.h"
+#include "wsp_utils.h"
+
+using namespace kuzu::binder;
+using namespace kuzu::common;
+
+namespace kuzu {
+namespace function {
+
+template<typename T>
+class AWSPPathsEdgeCompute : public EdgeCompute {
+public:
+    explicit AWSPPathsEdgeCompute(BFSGraph& bfsGraph) : bfsGraph{bfsGraph} {
+        block = bfsGraph.addNewBlock();
+    }
+
+    std::vector<nodeID_t> edgeCompute(nodeID_t boundNodeID, graph::NbrScanState::Chunk& chunk,
+        bool fwdEdge) override {
+        std::vector<nodeID_t> result;
+        chunk.forEach<T>([&](auto nbrNodeID, auto edgeID, auto weight) {
+            if (!block->hasSpace()) {
+                block = bfsGraph.addNewBlock();
+            }
+            if (bfsGraph.tryAddParentWithWeight(boundNodeID, edgeID, nbrNodeID, fwdEdge,
+                    (double)weight, block)) {
+                result.push_back(nbrNodeID);
+            }
+        });
+        return result;
+    }
+
+    std::unique_ptr<EdgeCompute> copy() override {
+        return std::make_unique<AWSPPathsEdgeCompute<T>>(bfsGraph);
+    }
+
+private:
+    BFSGraph& bfsGraph;
+    ObjectBlock<ParentList>* block = nullptr;
+};
+
+class AWSPPathsOutputWriter : public PathsOutputWriter {
+public:
+    AWSPPathsOutputWriter(main::ClientContext* context,
+        processor::NodeOffsetMaskMap* outputNodeMask, common::nodeID_t sourceNodeID,
+        PathsOutputWriterInfo info, BFSGraph& bfsGraph)
+        : PathsOutputWriter{context, outputNodeMask, sourceNodeID, info, bfsGraph} {
+        costVector = createVector(LogicalType::DOUBLE(), context->getMemoryManager());
+    }
+
+    void write(processor::FactorizedTable& fTable, common::nodeID_t dstNodeID,
```

Comment:
The control logic here looks exactly the same as `PathsOutputWriter::dfsFast` except a few control conditions. Can you refactor them together? Also, I'm curious if the fast and slow path applies to awsp too.

Reply:
As we discussed we will solve this gradually.